modules/REST/V2/Languages.php

```diff
@@ -0,0 +1,737 @@
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::EDITABLE ),
+				),
+				array(
+					'methods'             => WP_REST_Server::DELETABLE,
+					'callback'            => array( $this, 'delete_item' ),
+					'permission_callback' => array( $this, 'delete_item_permissions_check' ),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Retrieves a collection of languages.
+	 *
+	 * @since 3.7
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 *
+	 * @phpstan-template T of Fields
+	 * @phpstan-param WP_REST_Request<T> $request
+	 */
+	public function get_items( $request ) {
+		$languages = array();
+
+		foreach ( $this->model->languages_model->get_languages_list() as $language ) {
+			$languages[] = $this->prepare_item_for_response( $language, $request );
+		}
+
+		return rest_ensure_response( $languages );
+	}
+
+	/**
+	 * Creates one language from the collection.
+	 *
+	 * @since 3.7
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 *
+	 * @phpstan-template T of Fields
+	 * @phpstan-param WP_REST_Request<T> $request
+	 */
+	public function create_item( $request ) {
+		if ( ! empty( $request['id'] ) ) {
+			return new WP_Error(
+				'rest_exists',
+				__( 'Cannot create existing language.', 'polylang' ),
+				array( 'status' => 400 )
+			);
+		}
+
+		/**
+		 * $request: name, code, locale, direction, order,      flag
+		 * $args:    name, slug, locale, rtl,       term_group, flag? (from add_language())
+		 */
+		if ( ! isset( $request['name'], $request['code'], $request['direction'], $request['flag'] ) ) {
+			$languages = include POLYLANG_DIR . '/settings/languages.php';
+
+			if ( empty( $languages[ $request['locale'] ] ) ) {
+				return new WP_Error(
+					'rest_invalid_locale',
+					__( 'The locale is invalid.', 'polylang' ),
+					array( 'status' => 400 )
+				);
+			}
+
+			$language = $languages[ $request['locale'] ];
+			$defaults = array(
+				'name'       => $language['name'],
+				'slug'       => $language['code'],
+				'rtl'        => (int) ( 'rtl' === $language['dir'] ),
+				'flag'       => $language['flag'],
+				'term_group' => 0,
+			);
+		} else {
+			$defaults = array(
+				'term_group' => 0,
+			);
+		}
+
+		$prepared = (array) $this->prepare_item_for_database( $request );
+		$prepared = array_merge( $defaults, $prepared );
+
+		/**
+		 * @phpstan-var array{
+		 *     locale: non-empty-string,
+		 *     name: string,
+		 *     slug: string,
+		 *     rtl: 0|1,
+		 *     flag: string,
+		 *     term_group: 0|1
+		 * } $prepared
+		 */
+		$result = $this->model->languages_model->add( $prepared );
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
```

Comment:
Should we take care of error codes here, so it's more suitable for REST.
Like you've done in other places with `rest_invalid_locale` for instance and a proper status.

Reply:
OK for the status code.
However for the error code I don't think it's a good idea, we would loose information. Or maybe we could prefix the code with `rest_`? I'm not sure this worth it.
`add()` can return these errors:
```
'pll_invalid_locale'  => 'Enter a valid WordPress locale'
'pll_invalid_slug'    => 'The language code contains invalid characters'
'pll_non_unique_slug' => 'The language code must be unique'
'pll_invalid_name'    => 'The language must have a name'
'pll_invalid_flag'    => 'The flag does not exist'
'pll_add_language'    => 'Impossible to add the language.' (wp_insert_term() failed)
```

What do you think, replace `pll_` by `rest_`?

Reply:
Done in [8610884](https://github.com/polylang/polylang/pull/1505/commits/8610884aa391bcd85d906bc6c16f51663210afc5).